pkg/distributor/otel.go

```diff
@@ -109,43 +111,60 @@ func OTLPHandler(
 
 		pushErr := push(ctx, req)
 		if pushErr == nil {
-			// Push was successful, but OTLP converter left out some samples. We let the client know about it by replying with 4xx (and an insight log).
 			if otlpErr := otlpConverter.Err(); otlpErr != nil {
+				// Push was successful, but OTLP converter left out some samples. We let the client know about it by replying with 4xx (and an insight log).
 				pushErr = httpgrpc.Error(http.StatusBadRequest, otlpErr.Error())
-			}
-		}
-		if pushErr != nil {
-			if errors.Is(pushErr, context.Canceled) {
-				level.Warn(logger).Log("msg", "push request canceled", "err", pushErr)
-				writeErrorToHTTPResponseBody(r.Context(), w, statusClientClosedRequest, codes.Canceled, "push request context canceled", logger)
-				return
-			}
-			var (
-				httpCode int
-				grpcCode codes.Code
-				errorMsg string
-			)
-			if st, ok := grpcutil.ErrorToStatus(pushErr); ok {
-				// This code is needed for a correct handling of errors returned by the supplier function.
-				// These errors are created by using the httpgrpc package.
-				httpCode = httpRetryableToOTLPRetryable(int(st.Code()))
-				grpcCode = st.Code()
-				errorMsg = st.Message()
 			} else {
-				grpcCode, httpCode = toOtlpGRPCHTTPStatus(pushErr)
-				errorMsg = pushErr.Error()
-			}
-			if httpCode != 202 {
-				// This error message is consistent with error message in Prometheus remote-write handler, and ingester's ingest-storage pushToStorage method.
-				msgs := []interface{}{"msg", "detected an error while ingesting OTLP metrics request (the request may have been partially ingested)", "httpCode", httpCode, "err", pushErr}
-				if httpCode/100 == 4 {
-					msgs = append(msgs, "insight", true)
+				// Respond as per spec:
+				// https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#otlphttp-response.
+				var expResp colmetricpb.ExportMetricsServiceResponse
+				contentType := r.Header.Get("Content-Type")
+				if contentType == jsonContentType {
+					w.Header().Set("Content-Type", contentType)
+					data, _ := json.Marshal(&expResp)
+					_, _ = w.Write([]byte(data))
+				} else if contentType == pbContentType {
+					w.Header().Set("Content-Type", contentType)
+					data, _ := proto.Marshal(&expResp)
+					_, _ = w.Write(data)
+				} else {
+					level.Warn(logger).Log("msg", "unrecognized Content-Type request header", "content_type", contentType)
 				}
```

Comment:
I think we could remove the if statement for the protobuf case, and just assume if it's not json then the content-type is protobuf, similar to what happens for the error case. The request content type is checked in the function returned for `newOTLPParser` so we can be sure it's only json or protobuf by the time we're returning a response

```suggestion
					_, _ = w.Write([]byte(data))
				} else {
					w.Header().Set("Content-Type", contentType)
					data, _ := proto.Marshal(&expResp)
					_, _ = w.Write(data)
				}
```

If we do end up keeping the unrecogised content-type case though, we should make that return an error.

Reply:
Thanks, good point.